### ISSUE TITLE
Support advanced indexing from torch

### ISSUE THREAD

**Issue by Alphaharrius** [Alphaharrius/PyHilbert#59]
# Descriptions
The advanced indexing for a ` torch.Tensor` is triggered when a ` torch.Tensor` variable `i` is passed as a index for `tensor[i]`. The resulting operation becomes a gathering operation of the tensor.

This feature is required for band extraction by conditioning via the `torch.where` which currently our tensor API does not support.

# TODOs
- Enforce `StateSpace` elements to have `dim=1` and use `int` instead of `slice` in the structure
- Support advanced indexing such that a tensor `i` with dimensions `(A, B)` used as `tensor[i]` returns a tensor with shape `(A, B)`; in higher rank tensors the broadcasting rule much match with torch `ret[a, b, ...] = input[i[a, b, ...], j[a, b, ...], ...]`
- Implement `Tensor.where` and `Tensor.nonzero`
